akka_microservices/ibex35Microservice/src/main/scala/com/tapiporla/microservices/retrievers/common/ElasticDAO.scala

```diff
@@ -0,0 +1,174 @@
+
+}
+
+
+trait ElasticDAO extends Actor with ActorLogging with Stash{
+
+  def indexCreation: CreateIndexDefinition //Index to be created (if already exists we don't check if the mappings are similar)
+
+  self ! InitElasticDAO
+
+  override def receive = initial
+
+  def initial: Receive = {
+
+    case InitElasticDAO =>
+      client.execute(indexCreation) map { _ =>
+        ReadyToListen
+      } recover {
+        case CausedBy(ex: ResourceAlreadyExistsException) => //TODO: Change to see how to check if index already exists
+          ReadyToListen
+        case e: Exception =>
+          ProblemConnectingWithES(e)
+
+      } pipeTo (self)
+
+
+    case ProblemConnectingWithES(e) =>
+      log.error(s"Error initializing ElasticSearch DAO ${self.path.name} :$e. Trying again in 30 seconds")
+      context.system.scheduler.scheduleOnce(30 seconds, self, InitElasticDAO)
+
+
+    case ReadyToListen =>
+      unstashAll()
+      context.become(readyToProcess)
+
+    case _ =>
+      stash()
+  }
+
+
+  def readyToProcess: Receive = {
+
+    case SaveInIndex(index, typeName, jsonDocs) =>
+      client.execute {
+        bulk (
+          jsonDocs.map(x => indexInto(index / typeName).doc(x.json))
+        ) refresh(RefreshPolicy.WAIT_UNTIL)
+      } map { _=>
+        log.info(s"Completed to save in ES index $index type $typeName: ${jsonDocs.length} docs")
+        DataSavedConfirmation(index, typeName, jsonDocs)
+      } recover {
+        case e: Exception =>
+          log.error(s"Impossible to save in index $index $typeName ${jsonDocs.length} docs: $e")
+          ErrorSavingData(e, index, typeName, jsonDocs)
+      } pipeTo(sender)
+
+    case rq @ RetrieveAllFromIndexSorted(index, typeName, where, sort, limitValue) =>
+      client.execute {
+        where map { whereQuery =>
+          search(index / typeName) query whereQuery sortBy sort limit (limitValue.getOrElse(0))
+        } getOrElse
+          search(index / typeName) sortBy sort limit (limitValue.getOrElse(0))
+
+      } map { result =>
+        log.info(s"Retrieved from ES ${result.hits.length} elements")
+        DataRetrieved(index, typeName, result, rq)
+      } recover {
+        case e: Exception =>
+          log.error(s"Impossible to retrieve from index $index $typeName all sorted: $e")
+          ErrorRetrievingData(e, index, typeName, rq)
+      } pipeTo(sender)
+
+    case rq @ DeleteFromIndex(index, typeName, where) =>
+      client.execute {
+
+        where map { whereQuery =>
+          deleteIn(index / typeName).by(whereQuery)
+        } getOrElse
+          deleteIn(index / typeName).by(matchAllQuery())
+
+      } map { _ =>
+        log.info(s"Deleted confirmation in ES $index / $typeName")
+        DeleteConfirmation(index, typeName, rq)
+      } recover {
+        case e: Exception =>
+          log.error(s"Impossible to delete in index $index $typeName: $e")
```

Comment:
If you're not handling a concrete exception, would it be useful to see the exception trace?
```scala
log.error(s"Impossible to delete in index $index $typeName", e)
```

Reply:
Totally agree. Thanks

Reply:
Well, I changed the exceptions to be passed as parameter in log (which is correct).
Doing as before, (with $e) you don't see the stack trace (now yes), and for me is useful, because at the scratch of the deployment, may be you want to see if all is going OK (Exceptions are always welcome :P )